src/runtime/objmodel.cpp

```diff
@@ -405,6 +412,54 @@ void BoxedClass::finishInitialization() {
     commonClassSetup(this);
 }
 
+LocalShape BoxedClass::computeLocalShape() {
+    if (local_shape.isUninitialized()) {
+        std::vector<std::pair<llvm::StringRef, int>> attrs;
+
+        // iterate over all class attributes, storing the pairs of name, offset into attrs
+        auto hcls = getHCAttrsPtr()->hcls;
+        llvm::StringMap<int>::const_iterator it;
+        for (it = hcls->getStrAttrOffsets().begin(); it != hcls->getStrAttrOffsets().end(); it++) {
+            attrs.emplace_back(it->first(), it->second);
+        }
+        // sort the vector to hopefully give a deterministic ordering
+        std::sort(attrs.begin(), attrs.end(), [](std::pair<llvm::StringRef, int> a, std::pair<llvm::StringRef, int> b) {
+            return a.second < b.second;
+        });
+
+        // compute a hash of all attribute names
+        SHA256OStream hash_stream;
+        for (auto& s : attrs) {
+            hash_stream << s.first;
+        }
+        uint64_t lshape[4];
+        hash_stream.getHash(lshape);
+        local_shape = LocalShape(lshape[0], lshape[1], lshape[2], lshape[3]);
+    }
+
+    return local_shape;
+}
+
+Shape* BoxedClass::computeTotalShape() {
+    if (total_shape)
+        return total_shape;
+
+    Shape* shape = Shape::root_shape;
+    auto mro = static_cast<BoxedTuple*>(this->tp_mro);
+
+    assert(this == mro->elts[0]);
+
+    for (int i = mro->ob_size - 1; i >= 0; i--) {
+        BoxedClass* b = static_cast<BoxedClass*>(mro->elts[i]);
+
+        LocalShape b_shape = computeLocalShape();
```

Comment:
I'm not sure the elements of the mro are guaranteed to be BoxedClasses (I think you can simultaneously inherit from old-style and new-style classes) -- if it is the case, we should add an assert to that effect (`assert(isSubclass(mro->elts[i]->cls, type_cls)`).  If not, maybe these shape calculations don't need to be specific to BoxedClasses?  Though hopefully we would be able to do that without increasing the object size.